src/rows.cpp

```diff
@@ -120,124 +237,300 @@ SEXP process_slices(List& results, const List& slicers,
+    break;
+  }
+}
+
+inline void SlicesResults::add_colnames(List& out, std::string& output_colname) {
+  CharacterVector out_names = no_init(n_cols_);
+  if (n_labels_ > 0) {
+    CharacterVector slicing_cols_names = slicing_cols_.names();
+    std::copy(slicing_cols_names.begin(), slicing_cols_names.end(), out_names.begin());
+  }
+
+  switch (collation_) {
+  case rows:
+    add_colnames<rows>(out, output_colname, out_names);
+    break;
+  case cols:
+    add_colnames<cols>(out, output_colname, out_names);
+    break;
+  case list:
+    out_names[n_labels_] = output_colname;
+    break;
   }
 
   out.names() = out_names;
-  return as_data_frame(out);
 }
 
 
-// [[Rcpp::export]]
-SEXP by_slice_impl(const List& data, const SEXP fun, SEXP dots,
-                   int include_labels, const Environment& calling_env) {
-  ListOf<Symbol> slicers_symbols(data.attr("vars"));
-  int n_slicers(slicers_symbols.size());
+inline List SlicesResults::out(std::string output_colname) {
+  List out = no_init(n_cols_);
+
+  add_output(out);
+  add_labels(out);
+  add_colnames(out, output_colname);
+
+  return as_data_frame(out);
+}
 
-  CharacterVector slicers_names(n_slicers);
-  for (int i = 0; i != n_slicers; ++i) {
-    slicers_names[i] = slicers_symbols[i].c_str();
-  }
-  CharacterVector col_names = data.attr("names");
+
+List process_slices(List results, Environment env) {
+  SlicesResults slices_results(results, env);
+  CharacterVector output_colname = env[".to"];
+  return slices_results.out((std::string) output_colname[0]);
+}
 
 
-  // Evaluating a R call from C/Rcpp causes a language inlining of all
-  // arguments. This has a performance cost in some situations
-  // involving match.call(), and maybe in other cases as well. To
-  // avoid those, we store the slice in an environment.
-  Environment shadow_env(calling_env);
-  SET_TYPEOF(dots, LANGSXP);
-  dots = CDR(dots);  // car is the list that was used to capture dots
-  Shield<SEXP> lang_call(Rf_lcons(fun, Rf_lcons(Rf_install("shadowed"), dots)));
+SEXP subset_slices(const List data) {
+  ListOf<IntegerVector> indices(data.attr("indices"));
+  int n_slices = indices.size();
 
+  if (indices.size() == 0) {
+    Rf_error("Internal error: data not grouped");
+  }
 
-  // Remove slicing variables before subsetting rows
-  List results = subset_slices(data[!in(col_names, slicers_names)]);
+  CharacterVector classes = CharacterVector::create("tbl_df", "data.frame");
+  dplyr::DataFrameSubsetVisitors visitors(data);
 
-  for (int i = 0; i < results.size(); ++i) {
-    shadow_env.assign("shadowed", results[i]);
-    results[i] = Rcpp_eval(lang_call, shadow_env);
+  List out = no_init(n_slices);
+  for (int i = 0; i < n_slices; ++i) {
+    out[i] = visitors.subset(indices[i], classes);
   }
 
-  return process_slices(results, data[slicers_names],
-                        data.attr("labels"), include_labels);
+  return out;
 }
 
+extern "C" SEXP by_slice_impl(SEXP env, SEXP d_name_, SEXP f_name_) {
```

Comment:
It uses `R_ToplevelExec()` under the hood which ignores any pre-attached handlers. E.g. something like this would break:

```
withCallingHandlers(
    purrr::invoke_rows(data, function() message("Hello!"))
    message = function(m) {
        print("Caught message!")
    }
)
```